python/ray/data/impl/plan.py

```diff
@@ -293,20 +335,32 @@ def _optimize(self) -> Tuple[BlockList, DatasetStats, List[Stage]]:
             self._last_optimized_stages = stages
         return blocks, stats, stages
 
-    def _get_source_blocks(self) -> Tuple[BlockList, DatasetStats]:
-        """Get the source blocks (and corresponding stats) for plan execution.
+    def _get_source_blocks_and_stages(
+        self,
+    ) -> Tuple[BlockList, DatasetStats, List[Stage]]:
+        """Get the source blocks, corresponding stats, and the stages for plan
+        execution.
 
-        If a computed snapshot exists, return the snapshot blocks and stats; otherwise,
-        return the input blocks and stats that the plan was created with.
+        If a computed snapshot exists and has not been cleared, return the snapshot
```

Comment:
In retrospective, I think this essentially is (and could probably be called) "watermark" (of plan execution) - and then we can unify the two cases, be it input blocks (watermark at 0) or snapshot blocks (watermark at position N >= 1). (No need to change in this PR)

Reply:
Hmm I'm used to "watermark" being used for streaming systems in terms of handling "late" data, can you point me to where it's used in this context? I also feel like snapshot is a pretty accurate name, but happy to discuss alternatives.

Reply:
I think 'snapshot' as something that won't change whereas here we are tracking progress of a moving cursor. But I'm closing this thread since it's not public API (and not ask for a change in this PR), we can chat separately.